--- gym_donkeycar/envs/donkey_env.py (upstream 4ea6704)
+++ gym_donkeycar/envs/donkey_env.py (ours)
@@ -30,13 +30,15 @@
         ("start_delay", 5.0),
         ("max_cte", 8.0),
         ("frame_skip", 1),
-        ("cam_resolution", (120, 160, 3)),
+        #("cam_resolution", (120, 160, 3)),
+        ("cam_resolution", (64, 64, 3)),
         ("log_level", logging.INFO),
         ("host", "localhost"),
         ("port", 9091),
         ("steer_limit", 1.0),
         ("throttle_min", 0.0),
         ("throttle_max", 1.0),
+        ("headless", 1),
     ]
 
     for key, val in defaults:
@@ -50,7 +52,7 @@
     OpenAI Gym Environment for Donkey
 
     :param level: name of the level to load
-    :param conf: configuration dictionary
+    :param conf: configuration dictionaryf
     """
 
     metadata = {"render.modes": ["human", "rgb_array"]}
